src/main/java/com/segment/analytics/android/integrations/intercom/IntercomIntegration.java

```diff
@@ -43,29 +50,211 @@ public String key() {
         }
       };
 
+  private final Intercom intercom;
   private static final String INTERCOM_KEY = "Intercom";
   private final Logger logger;
 
-  public IntercomIntegration(Application application, ValueMap settings, Logger logger) {
-    String apiKey = settings.getString("apiKey");
+  // Intercom common specced attributes
+  private static final String NAME = "name";
+  private static final String CREATED_AT = "createdAt";
+  private static final String COMPANY = "company";
+  private static final String PRICE = "price";
+  private static final String CURRENCY = "currency";
+  private static final String AMOUNT = "amount";
+
+  // Intercom specced user attributes
+  private static final String EMAIL = "email";
+  private static final String PHONE = "phone";
+  private static final String LANGUAGE_OVERRIDE = "languageOverride";
+  private static final String UNSUBSCRIBED_FROM_EMAILS = "unsubscribedFromEmails";
+
+  // Intercom specced group attributes
+  private static final String MONTHLY_SPEND = "monthlySpend";
+  private static final String PLAN = "plan";
+
+  public interface Provider {
+
+    Intercom get();
+
+    Provider REAL =
+        new Provider() {
+          @Override
+          public Intercom get() {
+            return Intercom.client();
+          }
+        };
+  }
+
+  public IntercomIntegration(
+      Provider provider, Application application, ValueMap settings, Logger logger) {
+    String mobileApiKey = settings.getString("mobileApiKey");
     String appId = settings.getString("appId");
 
-    Intercom.initialize(application, apiKey, appId);
+    Intercom.initialize(application, mobileApiKey, appId);
+    this.intercom = provider.get();
     this.logger = logger;
   }
 
   @Override
   public void identify(IdentifyPayload identify) {
     super.identify(identify);
+
+    String userId = identify.userId();
+
+    if (isNullOrEmpty(userId)) {
+      intercom.registerUnidentifiedUser();
+      logger.verbose("Intercom.client().registerUnidentifiedUser()");
+    } else {
+      Registration registration = Registration.create().withUserId(userId);
+      intercom.registerIdentifiedUser(registration);
+      logger.verbose("Intercom.client().registerIdentifiedUser(registration)");
+    }
+
+    Map<String, Object> intercomOptions = identify.integrations().getValueMap("Intercom");
+    if (!isNullOrEmpty(intercomOptions)
+        && !isNullOrEmpty(String.valueOf(intercomOptions.get("userHash")))) {
+      intercom.setUserHash(String.valueOf(intercomOptions.get("userHash")));
+    }
+
+    Traits traits = identify.traits();
+    if (!isNullOrEmpty(traits) && (!isNullOrEmpty(intercomOptions))) {
+      setUserAttributes(traits, intercomOptions);
+      return;
+    }
+    setUserAttributes(traits, null);
```

Comment:
what if `intercomOptions` has values but `traits` do not?

Reply:
Traits will never be empty because the Android lib automatically adds userId and anonymousId as traits. We remove these in the setUserAttributes method after we make a copy of the immutable traits map.